### PR TITLE
this commit resolve second part of #12862

### PR DESCRIPTION
**Pull request by rahulsangwn**

### CODE REVIEW

File: binr/rax2/rax2.c [hunk 1]
```diff
@@ -228,8 +229,60 @@ static int rax(char *str, int len, int last) {
 		return true;
 	}
 	if (flags & (1 << 2)) { // -S
-		for (i = 0; i < len; i++) {
-			printf ("%02x", (ut8) str[i]);
+		for (int i = 0; i < len; i++) {
+			if (str[i] == '\\') {
+         
+				++i;
+				if (str[i] == 'n') printf("%02x", 10);
```
Comment (radare) [inline binr/rax2/rax2.c]: use a switch table and refactor this into a separate function that returns the dec value and just use a single printf statement

File: binr/rax2/rax2.c [hunk 2]
```diff
@@ -228,8 +229,60 @@ static int rax(char *str, int len, int last) {
...
+				else if (str[i] == '\0') printf("%02x", 0);
+				else if (str[i] == 'a') printf("%02x", 7);
+				else if (str[i] == 'b') printf("%02x", 8);
```
Comment (radare) [inline binr/rax2/rax2.c]: add a space before the  opening parenthesis (

File: binr/rax2/rax2.c [hunk 3]
```diff
@@ -228,8 +229,60 @@ static int rax(char *str, int len, int last) {
...
 	}
 	if (flags & (1 << 2)) { // -S
-		for (i = 0; i < len; i++) {
-			printf ("%02x", (ut8) str[i]);
+		for (int i = 0; i < len; i++) {
+			if (str[i] == '\\') {
+         
+				++i;
+				if (str[i] == 'n') printf("%02x", 10);
+				else if (str[i] == '\0') printf("%02x", 0);
```
Comment (radare) [inline binr/rax2/rax2.c]: use {} braces around the if statements, even if its just 1 line

File: binr/rax2/rax2.c [hunk 4]
```diff
@@ -228,8 +229,60 @@ static int rax(char *str, int len, int last) {
...
+				else if (str[i] == 't') printf("%02x", 9);
+				else if (str[i] == 'v') printf("%02x", 11);
+				else if (str[i] == '\\') printf("%02x", 92);
+				else if (str[i] == '\'') printf("%02x", 39);
+				else if (str[i] == '\"') printf("%02x", 34);
+				else if (str[i] == '\?') printf("%02x", 63);
+				else if (str[i] == 'x') {
+					i++;
+					while (IS_HEXCHAR (str[i])) {
+						printf("%c", tolower (str[i]));
```
Comment (radare) [inline binr/rax2/rax2.c]: missing space before (

File: binr/rax2/rax2.c [hunk 5]
```diff
@@ -228,8 +229,60 @@ static int rax(char *str, int len, int last) {
...
+						i++;
+					}
+					i--; 
+				} else if (IS_OCTAL (str[i])) {
+					if (!IS_OCTAL (str[i+1])) {
+						printf ("0%c", str[i]);
+					} else if ((IS_OCTAL (str[i+1])) && !IS_OCTAL (str[i+2])) {
+						char octal_str[3], *modified_str;
+						strncpy (octal_str, str+i, 2);
+						i += 1;
```
Comment (radare) [inline binr/rax2/rax2.c]: use i++

File: binr/rax2/rax2.c [hunk 6]
```diff
@@ -228,8 +229,60 @@ static int rax(char *str, int len, int last) {
...
+				else if (str[i] == '\"') printf("%02x", 34);
+				else if (str[i] == '\?') printf("%02x", 63);
+				else if (str[i] == 'x') {
+					i++;
+					while (IS_HEXCHAR (str[i])) {
+						printf("%c", tolower (str[i]));
+						i++;
+					}
+					i--; 
+				} else if (IS_OCTAL (str[i])) {
```
Comment (radare) [inline binr/rax2/rax2.c]: so many IS_OCTAL nested conditions, i dont see why is-octal should be checked for several bytes

File: binr/rax2/rax2.c [hunk 7]
```diff
@@ -3,6 +3,7 @@
 #include "../blob/version.c"
 #include <r_util.h>
 #include <r_util/r_print.h>
+#include <ctype.h>
```
[the remark this answers is not in the document]
Comment (rahulsangwn) [reply binr/rax2/rax2.c]: for tolower()

**Comment by radare** [comment]
ctype.h is alreay included in here, no need to include it again

$ git grep ctype.h libr/include/
libr/include/r_types_base.h:#include <ctype.h>

**Comment by rahulsangwn** [comment]
i made the changes @radare, but travis build failed again. i don't know what cause this and how i can fix it? please point little bit

**Comment by radare** [comment]
now there are 63 tests that are not passing, and no those tests are correct in  master

File: binr/rax2/rax2.c [hunk 8]
```diff
@@ -75,6 +75,24 @@ static void print_ascii_table() {
 	printf("%s", ret_ascii_table());
 }
 
+static int escape_to_dec(int i, char *str) {
+	switch (str[i]) {
+		case 'n': return 10; break;
```
Comment (radare) [inline binr/rax2/rax2.c]: case must be indented at the same level as the switch and there's no need to break if you return

File: binr/rax2/rax2.c [hunk 9]
```diff
@@ -75,6 +75,24 @@ static void print_ascii_table() {
 	printf("%s", ret_ascii_table());
 }
 
+static int escape_to_dec(int i, char *str) {
+	switch (str[i]) {
```
Comment (radare) [inline binr/rax2/rax2.c]: i think the index should be handled by the caller, not the function otherweise you have a boundary issue in here. so just switch(ch) {

File: binr/rax2/rax2.c [hunk 10]
```diff
@@ -228,8 +246,44 @@ static int rax(char *str, int len, int last) {
...
 	}
 	if (flags & (1 << 2)) { // -S
+		int i, dec_val;
 		for (i = 0; i < len; i++) {
-			printf ("%02x", (ut8) str[i]);
+			if (str[i] == '\\') {
+         		++i;
+				if ((dec_val = escape_to_dec (i, str)) != 2) {
+					printf ("%02x", dec_val);
+				}
```
Comment (radare) [inline binr/rax2/rax2.c]: else goes in the same place as the }

**Comment by rahulsangwn** [comment]
screenshot from 2019-01-31 21-41-09
I made the changes in `rax2`, now it correctly dumping the hash. But again travis build fails.

File: binr/ragg2/ragg2.c [hunk 11]
```diff
@@ -11,7 +11,7 @@
 static int usage(int v) {
 	printf ("Usage: ragg2 [-FOLsrxhvz] [-a arch] [-b bits] [-k os] [-o file] [-I path]\n"
 		"             [-i sc] [-e enc] [-B hex] [-c k=v] [-C file] [-p pad] [-q off]\n"
-		"             [-q off] [-dDw off:hex] file|f.asm|-\n");
+		"             [-f fmt] [-nN dword] [-dDw off:hex] file|f.asm|-\n");
```
[the remark this answers is not in the document]
Comment (radare) [reply binr/ragg2/ragg2.c]: this is probalby conflicting with current master

[the remark this answers is not in the document]
Comment (radare) [reply binr/ragg2/ragg2.c]: pls rebase

File: binr/rax2/rax2.c [hunk 12]
```diff
@@ -228,8 +246,44 @@ static int rax(char *str, int len, int last) {
 		return true;
 	}
 	if (flags & (1 << 2)) { // -S
+		int i, dec_val;
+		printf("%s\n", str);
 		for (i = 0; i < len; i++) {
-			printf ("%02x", (ut8) str[i]);
+			if (str[i] == '\\') {
+         		++i;
```
Comment (radare) [inline binr/rax2/rax2.c]: dont use preincrements when not necessary

File: binr/rax2/rax2.c [hunk 13]
```diff
@@ -228,8 +246,44 @@ static int rax(char *str, int len, int last) {
 		return true;
 	}
 	if (flags & (1 << 2)) { // -S
+		int i, dec_val;
+		printf("%s\n", str);
```
Comment (radare) [inline binr/rax2/rax2.c]: add space before (

File: binr/rax2/rax2.c [hunk 14]
```diff
@@ -228,8 +246,44 @@ static int rax(char *str, int len, int last) {
...
+						modified_str = r_str_newf (((*octal_str != '0') ? "0%s" : ""), octal_str);
+						ut64 n = r_num_math (num, modified_str); 
+						printf ("%02x", (ut8) n);
+					} else {	// '\ooo'
+						char octal_str[4], *modified_str;
+						strncpy (octal_str, str + i, 3);
+						i += 2;;
+						modified_str = r_str_newf (((*octal_str != '0') ? "0%s" : ""), octal_str);
+						ut64 n = r_num_math (num, modified_str);
+						((ut16) n) > 255 ? (printf ("%03x", (ut16) n)) : (printf ("%02x", (ut16) n));
```
Comment (radare) [inline binr/rax2/rax2.c]: dont use ternary operator for calls, here you just want it to chagne the format string

File: binr/rax2/rax2.c [hunk 15]
```diff
@@ -228,8 +246,44 @@ static int rax(char *str, int len, int last) {
...
+						char octal_str[3], *modified_str;
+						strncpy (octal_str, str + i, 2);
+						i++;
+						modified_str = r_str_newf (((*octal_str != '0') ? "0%s" : ""), octal_str);
+						ut64 n = r_num_math (num, modified_str); 
+						printf ("%02x", (ut8) n);
+					} else {	// '\ooo'
+						char octal_str[4], *modified_str;
+						strncpy (octal_str, str + i, 3);
+						i += 2;;
```
Comment (radare) [inline binr/rax2/rax2.c]: double semicolon

**Comment by radare** [comment]
please rebase and provide a test case

**Comment by radare** [comment]
please resolve conflicts and rebase

**Comment by rahulsangwn** [comment]
Yes i m working on it, one case is causing problem...i will fix it soon

**Pull request closed without merging**